include/deal.II/lac/tensor_product_matrix.h

```diff
@@ -730,25 +730,29 @@ ::reinit_impl (MatrixArray &&mass_matrices_,
   this->derivative_matrix = derivative_matrix;
 
   constexpr unsigned int macro_size = VectorizedArray<Number>::n_array_elements;
-  const unsigned int nm_flat_size
-    = (size > 0)
-      ? (Utilities::fixed_int_power<size,dim>::value
-         * Utilities::fixed_int_power<size,dim>::value * macro_size)
-      : (Utilities::fixed_power<dim>(mass_matrix[0].n_rows())
-         * Utilities::fixed_power<dim>(mass_matrix[0].n_rows()) * macro_size);
-  const unsigned int n_flat_size
-    = (size > 0)
-      ? Utilities::fixed_int_power<size,dim>::value * macro_size
-      : Utilities::fixed_power<dim>(mass_matrix[0].n_rows()) * macro_size;
+  const auto n_rows_max = [&mass_matrix]()
+  {
+    std::size_t n_rows_max {size};
+    if (size == -1)
+      {
+        std::array<std::size_t,dim> n_rows;
+        for (unsigned int d = 0; d < dim; ++d)
+          n_rows[d] = mass_matrix[d].n_rows();
+        n_rows_max = *(std::max_element (n_rows.cbegin(),n_rows.cend()));
+      }
+    return n_rows_max;
+  };
+  const unsigned int nm_flat_size_max = n_rows_max() * n_rows_max() * macro_size;
```

Comment:
I would not make this a lambda (without argument) since you just count the components. You can simply do this inline for the variable `n_rows_max`.